### PR TITLE
Allow custom priority threshold

### LINKED ISSUE #329: Allow Ignoring Pods For Eviction
**Is your feature request related to a problem? Please describe.**

I would like to be able to tell the descheduler to ignore certain pods for eviction. Currently the descheduler provides several options for ignoring pods, but none of the current options fit my use case.

**Describe the solution you'd like**

**Option 1:**
I would like to be able to set the annotation `descheduler.alpha.kubernetes.io/evict: "false"` on a pod and have descheduler always ignore this pod. Descheduler should never evict this pod. Users would need to add the annotation `descheduler.alpha.kubernetes.io/evict: "true"` to inform the descheduler to always evict a pod. This would be a breaking change.

**Option 2:**
I would like descheduler to have an option to specify a PriorityClass threshold that allows excluding pods for eviction. This is similar to ignoring system and node critical pods, but the end user can specify a PriorityClass threshold instead of only allowing the hard coded system and node critical  PriorityClasses.

**Describe alternatives you've considered**

* Marking the pods as a critical pod
* Assigning local storage to the pod
* Setting a very restrictive PDB

Marking the pod as a [critical pod ](https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/)is the closest option that fits my use case, but the pods I want descheduler to ignore are not considered system or node critical. It is just a pod running a business application therefore I do not want to mark it as system or node critical.

**What version of descheduler are you using?**

descheduler version: v0.18.0

**Additional context**

The k8s cluster autoscaler currently supports a similar annotation for pods `"cluster-autoscaler.kubernetes.io/safe-to-evict": "false"`. See the [CA FAQ](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node) for details.

### PR DESCRIPTION
**Pull request by lixiang233**
Fixes #329 
Introduce a new configuration for each strategies which allows users to define a PriorityClass which is used by PodEvictor to check if a pod is evictable.

TODO:
- [x] Add PriorityClass to `StrategyParameters` and read and set it to PodEvictor
- [x] e2e tests for priority class
- [x] README
- [x] e2e tests for priority

### CODE REVIEW

**Comment by lixiang233** [comment]
/cc @ingvagabund  @seanmalloy 
/assign @damemi 
Please take a look, this pr is ready for review and will be ready for merge once I have updated README.

File: pkg/descheduler/strategies/duplicates.go [hunk 1]
```diff
@@ -44,6 +44,17 @@ func RemoveDuplicatePods(
...
+	var priorityClass string
+	if strategy.Params != nil {
+		priorityClass = strategy.Params.ThresholdPriorityClassName
+	}
+	priority, err := podEvictor.SetPriorityThresholdFromPriorityClass(ctx, client, priorityClass)
+	if err != nil {
+		klog.V(1).Infof("Error setting priority threshold, %#v", err)
+		return
+	}
+	klog.V(1).Infof("Running RemoveDuplicatePods with priority threshold %d", priority)
```
Comment (damemi) [inline pkg/descheduler/strategies/duplicates.go:56]: I think we should only log the threshold if it's being set to something that's not default.

File: pkg/descheduler/strategies/lownodeutilization.go [hunk 2]
```diff
@@ -53,6 +53,17 @@ const (
...
 func LowNodeUtilization(ctx context.Context, client clientset.Interface, strategy api.DeschedulerStrategy, nodes []*v1.Node, podEvictor *evictions.PodEvictor) {
+	var priorityClass string
+	if strategy.Params != nil {
+		priorityClass = strategy.Params.ThresholdPriorityClassName
+	}
+	priority, err := podEvictor.SetPriorityThresholdFromPriorityClass(ctx, client, priorityClass)
+	if err != nil {
+		klog.V(1).Infof("Error setting priority threshold, %#v", err)
+		return
+	}
```
Comment (damemi) [inline pkg/descheduler/strategies/lownodeutilization.go:64]: You could simplify this a bit like:
```go
if strategy.Params != nil {
  priority, err := podEvictor.SetPriorityThresholdFromPriorityClass(ctx, client, strategy.Params.ThresholdPriorityClassName)
  if err != nil {
    klog.V(1).Errorf("Error setting priority threshold: %+v", err)
    return
  }
  klog.V(1).Infof("Running LowNodeUtilization with priority threshold %d", priority)
}
```

Or just pass `*strategyParams` to `podEvictor.SetPriorityThresholdFromPriorityClass` and do the nil check there (and the log message below this if it's successfully set). That would clean up some of the copy paste in each strategy

File: pkg/descheduler/strategies/pod_lifetime.go [hunk 3]
```diff
@@ -49,6 +49,13 @@ func PodLifeTime(ctx context.Context, client clientset.Interface, strategy api.D
 		klog.V(1).Info(err)
 		return
 	}
+	priorityClass := strategy.Params.ThresholdPriorityClassName
+	priority, err := podEvictor.SetPriorityThresholdFromPriorityClass(ctx, client, priorityClass)
```
Comment (damemi) [inline pkg/descheduler/strategies/pod_lifetime.go:53]: any reason to not pass `strategy.Params.ThresholdPriorityClassName` directly to the function?

**Comment by damemi** [comment]
btw when your PR is ready for review in the future just remove the [WIP] and we'll know it's ready :)
We can /hold if you're waiting for pieces like README updates that will depend on feedback
/retitle Allow custom priority threshold
/hold

**Comment by lixiang233** [comment]
> We could also consider just adding this as a `filter` function 

@damemi  I prefer this approach, as `IsPodEvictableBasedOnPriority` will be called for every pod no matter it's in `IsEvictable` or the filter function, just moving `IsPodEvictableBasedOnPriority` out of `IsEvictable` won't affect the performance. and if we add it to `EvictPod`, we may do some useless calculations because some pods may not be evictable and we could filter out them before running the strategy.

At first I've thought of many ways to do this, given we'll set priority threshold every time a strategy starts and our strategies are running one by one, I think it doesn't matter if one strategy may affect others, besides I think we should do all `IsEvictable` checks together in one function, so I choose to make the threshold a param of `PodEvictor`. That's not a very good implementation, I'll reimplement it using the methods mentioned above and add another configuration to set the threshold value directly.

**Comment by ingvagabund** [comment]
I second what @damemi says.
For strategies which are not cluster wide (RemovePodsViolatingInterPodAntiAffinity, RemovePodsViolatingNodeAffinity, RemovePodsViolatingNodeTaints, ...), you can extend the filters so you don't bother with pods what have higher priority class than ThresholdPriorityClassName. In which case the pod evictor does not have to check the priority class at all.
For the remaining strategies (RemoveDuplicates, LowNodeUtilization), you can always check a pod's class right before you pass the pod to the evictor. So the pod evictor can stay intact.

**Comment by lixiang233** [comment]
As `All types of pods with the annotation descheduler.alpha.kubernetes.io/evict are evicted.` and this annotation is checked in `IsEvictable`, if I add the priority check as an individual filter function, and use them like:
```
podutil.WithFilter(func(pod *v1.Pod) bool {
      return podEvictor.IsEvictable(pod) && isPodEvictableByPriority(pod)
})
```
Although a pod has the annotation, if its priority is above priority threshold, it won't be evictable.

So if we want this annotation still work for priority filtering, I have some solutions:
1.add priority check to `IsEvictable` and refactor `IsEvictable` to have a parameter.
2.check this annotation again in `isPodEvictableByPriority`
3.make annotation check another filter func and call them like:
```
podutil.WithFilter(func(pod *v1.Pod) bool {
      return (podEvictor.IsEvictable(pod) && isPodEvictableByPriority(pod)) || isPodEvictableByAnnotation(pod)
})
```
I prefer 1, what do you think @damemi  @ingvagabund ?

**Comment by lixiang233** [comment]
I have implemented this
> 1.add priority check to `IsEvictable` and refactor `IsEvictable` to have a parameter.

and added `ThresholdPriority` to strategy's params.

File: pkg/utils/pod.go [hunk 4]
```diff
@@ -112,15 +112,12 @@ func IsCriticalPod(pod *v1.Pod) bool {
 	if IsMirrorPod(pod) {
 		return true
 	}
-	if pod.Spec.Priority != nil && IsCriticalPodBasedOnPriority(*pod.Spec.Priority) {
```
Comment (ingvagabund) [inline pkg/utils/pod.go:115]: This needs to stay here. You should never be allowed to evict critical pods.

Comment (damemi) [reply pkg/utils/pod.go:115]: Isn't this check basically replaced by IsPodEvictableBasedOnPriority?

**Comment by lixiang233** [comment]
I've moved `IsPodEvictableBasedOnPriority` to the `evictions` package and added e2e testcase for `ThresholdPriority`.

please take a look @damemi  @ingvagabund

File: pkg/utils/pod.go (hunk 4 above)
Comment (ingvagabund) [reply pkg/utils/pod.go:115]: Given the way it's implemented:

```
func IsPodEvictableBasedOnPriority(pod *v1.Pod, priority int32) bool {
	return pod.Spec.Priority == nil || *pod.Spec.Priority < priority
}
```

a user can choose any int32 value. Even the one greater than the system critical priority. Thus, if improperly configured the descheduler can evict even the critical pods. `return pod.Spec.Priority == nil || (*pod.Spec.Priority < priority && *pod.Spec.Priority < SystemCriticalPriority)` would do a bit better.

My point is `IsEvictable` function is supposed to be static. The hardline for deciding when a pod is evictable or not without any degree of flexibility. If `IsEvictable` decides a pod is not evictable, it is not. It's a safe guard condition that can not be changed from the outside (except for setting some true/false flags).

Comment (damemi) [reply pkg/utils/pod.go:115]: Good points, I did notice this change at first too but thought it was just being covered by the new code.

In that case, we should probably keep it the same as it was right?

Comment (damemi) [reply pkg/utils/pod.go:115]: @lixiang233 what do you think?

Comment (lixiang233) [reply pkg/utils/pod.go:115]: Agree, if someone wants to evict system-critical pods, he can add `descheduler.alpha.kubernetes.io/evict` annotation. I'll add `*pod.Spec.Priority < SystemCriticalPriority` to `IsPodEvictableBasedOnPriority`.

**Comment by lixiang233** [comment]
I've completed the change mentioned before and added a validation to the threshold. @damemi  @ingvagabund

File: pkg/utils/priority.go [hunk 5]
```diff
@@ -33,3 +39,37 @@ func PodMatchesTermsNamespaceAndSelector(pod *v1.Pod, namespaces sets.String, se
...
+// GetPriorityFromPriorityClass gets priority from the given priority class.
+// If no priority class is provided, it will return SystemCriticalPriority by default.
+func GetPriorityFromPriorityClass(ctx context.Context, client clientset.Interface, name string) (int32, error) {
+	if name != "" {
+		priorityClass, err := client.SchedulingV1().PriorityClasses().Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return 0, err
+		} else {
+			return priorityClass.Value, nil
+		}
```
Comment (seanmalloy) [inline pkg/utils/priority.go:52]:
```suggestion
		}
		return priorityClass.Value, nil
```

Comment (lixiang233) [reply pkg/utils/priority.go]: done.

**Pull request merged**